### ISSUE TITLE
Better codegen for literal 0 on x86

### ISSUE THREAD

**Issue by jyn514** [CraneStation/cranelift#897]
Take the following C program:

```c
int main(void) {
    return 0;
}
```

Clang and gcc both compile it to `xor eax, eax` (see https://godbolt.org/z/7q4_bq and https://godbolt.org/z/ADm8ZQ):

```
main:                                   # @main
        xor     eax, eax
        ret
```

Cranelift compiles it to `mov $0, eax`:

```
 600:   55                      push   %rbp
 601:   48 89 e5                mov    %rsp,%rbp
 604:   b8 00 00 00 00          mov    $0x0,%eax
 609:   5d                      pop    %rbp
 60a:   c3                      retq   
 ```

Compiling to `xor eax, eax` is beneficial because it uses fewer opcodes, see https://stackoverflow.com/questions/33666617/what-is-the-best-way-to-set-a-register-to-zero-in-x86-assembly-xor-mov-or-and and https://stackoverflow.com/questions/1396527/what-is-the-purpose-of-xoring-a-register-with-itself#1396552 for a detailed discussion.

<details><summary>Full reproduction with code</summary>

```
use std::fs::File;                                                                           
                                                                                             
use cranelift::codegen::ir::Function;                                                        
use cranelift::codegen::isa::CallConv;                                                       
use cranelift::prelude::*;                                                                   
use cranelift_faerie::{FaerieBackend, FaerieBuilder, FaerieProduct, FaerieTrapCollection};   
use cranelift_module::{Linkage, Module};                                                     
use target_lexicon::Triple;                                                                  
                                                                                             
fn main() {                                                                                  
    let mut builder = settings::builder();                                                   
    builder.enable("is_pic").unwrap();
    builder.set("opt_level", "best").unwrap();
    let flags = settings::Flags::new(builder);

    let isa = isa::lookup(Triple::host()).unwrap().finish(flags.clone());
    let builder = FaerieBuilder::new(
        isa,
        "<test>".into(),
        FaerieTrapCollection::Disabled,
        cranelift_module::default_libcall_names(),
    )
    .unwrap();
    let mut module = Module::<FaerieBackend>::new(builder);

    let params = Vec::new();
    let returns = vec![AbiParam::new(types::I32)];
    let signature = Signature {
        call_conv: CallConv::triple_default(&Triple::host()),
        params,
        returns,
    };
    let func_id = module
        .declare_function("<test function>", Linkage::Export, &signature)
        .unwrap();
    let mut func = Function::with_name_signature(ExternalName::user(0, 0), signature);

    let mut ctx = FunctionBuilderContext::new();
    let mut builder = FunctionBuilder::new(&mut func, &mut ctx);
    let func_start = builder.create_ebb();
    builder.append_ebb_params_for_function_params(func_start);
    builder.switch_to_block(func_start);

    // the actual code
    let zero = builder.ins().iconst(types::I32, 0);
    builder.ins().return_(&[zero]);

    codegen::verify_function(&func, &flags).unwrap();

    let mut ctx = codegen::Context::for_function(func);
    module.define_function(func_id, &mut ctx).unwrap();

    let product: FaerieProduct = module.finish();
    let file = File::create("a.out").unwrap();
    product.write(file).unwrap();
}
```

</details>

**Comment by iximeow** [comment]
i'll take this as a motivator to fix up https://github.com/CraneStation/cranelift/pull/766 :)

**Comment by sstangl** [comment]
We should also realize that the stack is unused and avoid saving rbp. That's on my list: it comes up a lot with simple functions.

**Comment by bnjbvr** [comment]
This was fixed by #766. The spurious stack save and restore could be handled as a different issue.
